### PR TITLE
API-19 update content api post

### PR DESCRIPTION
**Pull request by nebc-jasonfung**
- API-19 update content api post
- Update readme upload link

### CODE REVIEW

File: content-api-post.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ exports.handler = function (event, context, callback) {
   const data = JSON.parse(event.body);
   const decodedImage = Buffer.from(data.body, 'base64');
 
-  if (!decodedImage || !data.accountId || !data.type) {
+  if (!decodedImage || !data.accountId || !data.championFundId || !data.type) {
```
Comment (NoelLH) [inline content-api-post.js]: @tbg-jasonfung this doesn't seem right. What about uploads for non-funds?

File: Readme.md [hunk 2]
```diff
@@ -12,4 +12,4 @@ which provides the `POST` endpoint for the Big Give's user-generated content API
 
 * Zip up the *contents* of the folder (not the whole folder), including `node_modules/`. This
   need not include normally-hidden files like `.git*` metadata.
-* Upload [here](https://eu-west-2.console.aws.amazon.com/lambda/home?region=eu-west-2#/functions/content-api-post-sandbox?tab=graph).
+* Upload [here](https://eu-west-1.console.aws.amazon.com/lambda/home?region=eu-west-1#/functions).
```
Comment (NoelLH) [inline Readme.md]: Are you testing this with a temporary Lambda name? Might be good to keep this in explicitly so we don't mix up manual, temporary deployments with the Terraform-managed ones

**Review by NoelLH — CHANGES_REQUESTED** [review]

File: Readme.md (hunk 2 above)
Comment (nebc-jasonfung) [reply Readme.md]: Oh I see, so manual temporary ones I need to create a new function to test - I saw the page was empty and thought this was the wrong link, I've reverted the change @NoelLH

Comment (nebc-jasonfung) [reply Readme.md]: I've tried to create a temporary one, but I'm blocked with permissions error:
image

Comment (NoelLH) [reply Readme.md]: @tbg-jasonfung thinking about this more, as the change is not too drastic and we're now updating code that is live / used in Prod, it's probably better to get things through code review and then go straight to applying updates with Terraform in the normal, systematic way, with the staging `apply` and manual QA happening first. The Terraform user whose key you use will also have permissions to programmatically create and update Lambdas.

File: content-api-post.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ exports.handler = function (event, context, callback) {
   const data = JSON.parse(event.body);
   const decodedImage = Buffer.from(data.body, 'base64');
 
-  if (!decodedImage || !data.accountId || !data.type) {
+  if (!decodedImage || !(data.accountId && data.championFundId) || !data.type) {
```
Comment (NoelLH) [inline content-api-post.js]: @tbg-jasonfung this still doesn't look right to me. Think about the case where there is an account ID but no fund. Parenthesised like this you're still requiring both.

**Review by NoelLH — CHANGES_REQUESTED** [review]

File: content-api-post.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ exports.handler = function (event, context, callback) {
   const data = JSON.parse(event.body);
   const decodedImage = Buffer.from(data.body, 'base64');
 
-  if (!decodedImage || !data.accountId || !data.type) {
+  if (!decodedImage || (!data.accountId && !data.championFundId) || !data.type) {
```
Comment (nebc-jasonfung) [inline content-api-post.js:30]: @NoelLH This should now fail if both are empty, and will pass if either 1 is true - I'll apply the changes through the Terraform way once this is okay

File: content-api-post.js [hunk 5]
```diff
@@ -56,6 +60,9 @@ exports.handler = function (event, context, callback) {
         SalesforceAccountId: data.accountId,
       };
       // All remaining metadata keys are optional. We can't append `null`s as this is not a valid value for headers.
+      if (data.championFundId) {
```
Comment (NoelLH) [inline content-api-post.js:69]: @tbg-jasonfung  what about the path? This change on its own surely doesn't work - the most critical part (where the file lives) still relies on account ID.

Or at least, if path is to continue relying on account ID, then you need to keep account ID unconditionally required and send it from the client. I guess this is possible since a fund belongs to only one champion but you should think about whether it's expected, align the validation accordingly, and document it.

I've started a possible API contract update at https://app.swaggerhub.com/apis/Noel/TBG-Content/1.1.0 but I would suggest starting with this yourself first. Should help to clarify what we're changing before building it.

**Review by NoelLH — CHANGES_REQUESTED** [review]

File: content-api-post.js (hunk 5 above)
Comment (nebc-jasonfung) [reply content-api-post.js:69]: @NoelLH Thanks - I've kept the accountId and championFundId metadata dynamic, I think it'll be easier to have the Ids set the correct metadata instead of relying on account Id

File: content-api-post.js [hunk 6]
```diff
@@ -42,6 +42,10 @@ exports.handler = function (event, context, callback) {
 
   const maxSize = 2500;
 
+  // Assign the Account record or Champion Fund Id.
+  // We know at least one of these exist before reaching here
+  SalesforceId = data.accountId ? data.accountId : data.championFundId;
```
Comment (NoelLH) [inline content-api-post.js]: @tbg-jasonfung variable declarations should use a suitable ES6 keyword. If you're not reassigning it that's probably `const`. If you open this in any good IDE you should get a big red warning about this!

Variable name should start with a lowercase for convention with the rest of the file / normal JS style.

`SalesforceId` doesn't really explain what the ID is used for. I think it's reasonable if a bit messy to vary which ID is the primary object linking to the file, but we should probably call it `salesforcePathId` or similar to hint at why we're setting this ternary up in the first place. It probably also makes more sense to do it only when we're about to use it, e.g. just before declaring `path`, otherwise we're executing code for no reason.

File: content-api-post.js [hunk 7]
```diff
@@ -55,14 +59,18 @@ exports.handler = function (event, context, callback) {
...
-      const metadata = {
-        SalesforceAccountId: data.accountId,
-      };
-      // All remaining metadata keys are optional. We can't append `null`s as this is not a valid value for headers.
+      const path = `${SalesforceId}/${data.type}/${generatedName}`;
+      const metadata = {};
+
+      // Assign the dynamic SF Id to the correct metadata
+      if (data.accountId) {
+        metadata.SalesforceAccountId = SalesforceId;
```
Comment (NoelLH) [inline content-api-post.js]: I don't see any benefit in using the less descriptive, new variable here - we only need it for the S3 path setup. As it was before you could tell what the value is without having to grok another variable, and it does the same thing.

**Review by NoelLH — CHANGES_REQUESTED** [review]

File: content-api-post.js [hunk 8]
```diff
@@ -63,15 +59,17 @@ exports.handler = function (event, context, callback) {
     .toBuffer()
     .then(processedImage => {
       const generatedName = `${uuidv4()}.${mimeType.ext}`;
-      const path = `${SalesforceId}/${data.type}/${generatedName}`;
+      const salesforcePathId  = data.accountId ? data.accountId : data.championFundId;
```
Comment (NoelLH) [inline content-api-post.js]: @tbg-jasonfung weird spacing here

File: content-api-post.js [hunk 9]
```diff
@@ -27,10 +27,18 @@ exports.handler = function (event, context, callback) {
   const data = JSON.parse(event.body);
   const decodedImage = Buffer.from(data.body, 'base64');
 
-  if (!decodedImage || !data.accountId || !data.type) {
+  if (!decodedImage || (!data.accountId && !data.championFundId) || !data.type) {
     return fail('Missing required metadata', 400, callback);
   }
 
+  if (!data.accountId && data.championFundId && data.ccampaignId) {
```
Comment (NoelLH) [inline content-api-post.js]: @tbg-jasonfung what's the rationale for having all 3 parts here? Given the below checks we're not combining account + fund, shouldn't this just check fo campaign + fund?

Comment (nebc-jasonfung) [reply content-api-post.js]: @NoelLH Nicee, I've also removed the duplicate return fail message and bunged these checks all into 1 line

**Review by NoelLH — APPROVED** [review]
